ratatool-scalacheck/src/main/scala/com/spotify/ratatool/scalacheck/ProtoBufGenerator.scala

```diff
@@ -0,0 +1,189 @@
+
+  case class NullWriter()(cos: CodedOutputStream) extends Writer(0, Unit)(cos)
+  { override def write(): Unit = Unit }
+
+  case class RepeatedWriter(v: List[PartialWriter])(cos: CodedOutputStream)
+    extends Writer(0, v)(cos)
+  {
+    override def write(): Unit = v.foreach(pw => pw(cos).write())
+  }
+}
+
+object ProtoBufGeneratorOps extends ProtoBufGeneratorOps
+
+trait ProtoBufGeneratorOps {
+  import ProtoBufWriters._
+
+  private val cache: LoadingCache[Class[_], (Descriptors.Descriptor, Method)] = CacheBuilder
+    .newBuilder()
+    .build(new CacheLoader[Class[_], (Descriptors.Descriptor, Method)] {
+      override def load(key: Class[_]): (Descriptor, Method) = {
+        val desc = key.getMethod("getDescriptor").invoke(null).asInstanceOf[Descriptors.Descriptor]
+        val parseFn = key.getMethod("parseFrom", classOf[Array[Byte]])
+        (desc, parseFn)
+      }
+    })
+
+  /** Generate a ProtoBuf Message */
+  def protoBufOf[T <: AbstractMessage : ClassTag]: Gen[T] = {
+    val (desc, parseFn) = cache.get(implicitly[ClassTag[T]].runtimeClass)
+    val bytesGen = generate(desc)
+    bytesGen.map(parseFn.invoke(null, _).asInstanceOf[T])
+  }
+
+  private def generate(desc: Descriptors.Descriptor): Gen[Array[Byte]] = {
+    Gen
+      .sequence[List[PartialWriter], PartialWriter](desc.getFields.asScala.map(genField)).map { l =>
+      val baos = new ByteArrayOutputStream()
+      val cos = CodedOutputStream.newInstance(baos)
+      l.foreach(pw => pw(cos).write())
+      cos.flush()
+      baos.toByteArray
+    }
+  }
+
+  private def genPositiveInt: Gen[Int] = Gen.chooseNum(0, Int.MaxValue)
+  private def genPositiveLong: Gen[Long] = Gen.chooseNum(0L, Long.MaxValue)
+
+  // scalastyle:off cyclomatic.complexity method.length
+  private def genField(field: FieldDescriptor): Gen[PartialWriter] = {
+    val id = field.getNumber
+    def genV: Gen[PartialWriter] = field.getType match {
+      case Type.DOUBLE => Arbitrary.arbDouble.arbitrary.map(v => DoubleWriter(id, v))
+      case Type.FLOAT => Arbitrary.arbFloat.arbitrary.map(v => FloatWriter(id, v))
+      case Type.INT32 => Arbitrary.arbInt.arbitrary.map(v => Int32Writer(id, v))
+      case Type.INT64 => Arbitrary.arbLong.arbitrary.map(v => Int64Writer(id, v))
+      case Type.UINT32 => genPositiveInt.map(v => UInt32Writer(id, v))
+      case Type.UINT64 => genPositiveLong.map(v => UInt64Writer(id, v))
+      case Type.SINT32 => genPositiveInt.map(v => SInt32Writer(id, v))
+      case Type.SINT64 => genPositiveLong.map(v => SInt64Writer(id, v))
+      case Type.FIXED32 => genPositiveInt.map(v => Fixed32Writer(id, v))
+      case Type.FIXED64 => genPositiveLong.map(v => Fixed64Writer(id, v))
+      case Type.SFIXED32 => genPositiveInt.map(v => SFixed32Writer(id, v))
+      case Type.SFIXED64 => genPositiveLong.map(v => SFixed64Writer(id, v))
+      case Type.BOOL => Arbitrary.arbBool.arbitrary.map(v => BoolWriter(id, v))
+      case Type.STRING => Gen.choose(0, 40)
```

Comment:
what's this magic range for

should we add a comment to explain

Reply:
Mostly to mimic behaviour here https://github.com/spotify/ratatool/blob/master/ratatool/src/main/scala/com/spotify/ratatool/generators/ProtoBufGenerator.scala#L85

I'm not certain why this is limited to begin with though. @nevillelyh do you know?

Reply:
I'll move to an arbitrary string for now, can revert later if needed